### PR TITLE
Property Test for SignedMessage

### PR DESCRIPTION
**Pull request by jdjaustin**
**Summary of changes**
Changes introduced in this pull request:
- Use quickcheck to test our custom implementations of `SignedMessage`

**Reference issue to close (if applicable)**

Closes associated task in #1881 

**Other information and links**

N/A

### CODE REVIEW

**Comment by lemmih** [comment]
@jdjaustin All good? It's unfortunate that `serde_json::to_string` cannot be safely used without `SignedMessageJson` but it's out of our hands.

File: vm/message/Cargo.toml [hunk 1]
```diff
@@ -23,13 +23,18 @@ fvm               = "1.1"
...
+serde_json        = "1.0"
 
 [features]
 default = ["blst"]
 blst    = ["forest_crypto/blst", "fvm_shared/blst"]
 
 proofs = []
+
+[package.metadata.cargo-udeps.ignore]
+normal = ["quickcheck"]
```
Comment (lemmih) [inline vm/message/Cargo.toml:40]: Remove this.

File: vm/message/src/signed_message.rs [hunk 2]
```diff
@@ -124,6 +124,35 @@ impl Cbor for SignedMessage {
...
+        Ok(Signature::new_secp256k1(DUMMY_SIG.to_vec()))
+    }
+}
+
+// impl quickcheck::Arbitrary for dyn Signer {
+//     fn arbitrary(g: &mut quickcheck::Gen) -> Self {
+//         let dummy_signer = Signer::sign_bytes(_, g, Address).unwrap();
+//         dummy_signer
+//     }
+// }
```
Comment (lemmih) [inline vm/message/src/signed_message.rs:142]: Remove this.

File: vm/message/src/signed_message.rs [hunk 3]
```diff
@@ -124,6 +124,35 @@ impl Cbor for SignedMessage {
...
 
+const DUMMY_SIG: [u8; 1] = [0u8];
+
+//#[derive(Clone)]
+struct DummySigner;
+impl Signer for DummySigner {
+    fn sign_bytes(&self, _: &[u8], _: &Address) -> Result<Signature, anyhow::Error> {
+        Ok(Signature::new_secp256k1(DUMMY_SIG.to_vec()))
+    }
+}
```
Comment (lemmih) [inline vm/message/src/signed_message.rs:135]: Move this away from the top-level.

**Review by lemmih — CHANGES_REQUESTED** [review]

File: vm/message/Cargo.toml (hunk 1 above)
Comment (jdjaustin) [reply vm/message/Cargo.toml:40]: The PR was failing the unused dependencies check without this (`cargo udeps` apparently does not see dependencies used only in tests?).

Comment (lemmih) [reply vm/message/Cargo.toml:40]: If a dependency is only used for testing, it should be in `dev-dependencies`.

Comment (lemmih) [reply vm/message/Cargo.toml:40]: Suppressing warnings is a last resort and you should always ask in #fil-devs before doing it.

Comment (jdjaustin) [reply vm/message/Cargo.toml:40]: Ah but I see that one is not required.

Comment (jdjaustin) [reply vm/message/Cargo.toml]: I don't think we can put this one under `dev-dependencies` though. See the discussion under PR #1879 .

File: vm/message/src/signed_message.rs [hunk 4]
```diff
@@ -124,6 +126,25 @@ impl Cbor for SignedMessage {
     }
 }
 
+struct DummySigner;
+impl Signer for DummySigner {
+    fn sign_bytes(&self, _: &[u8], _: &Address) -> Result<Signature, anyhow::Error> {
+        Ok(Signature::new_secp256k1(DUMMY_SIG.to_vec()))
+    }
+}
```
Comment (lemmih) [inline vm/message/src/signed_message.rs:134]: You can move this into the body of `arbitrary`.

File: vm/message/Cargo.toml (hunk 1 above)
Comment (lemmih) [reply vm/message/Cargo.toml]: Have you tried?

Comment (jdjaustin) [reply vm/message/Cargo.toml]: I guess I didn't since I was just following the example. Seems to work!

File: vm/message/src/signed_message.rs [hunk 5]
```diff
@@ -124,6 +124,27 @@ impl Cbor for SignedMessage {
...
+            fn sign_bytes(&self, _: &[u8], _: &Address) -> Result<Signature, anyhow::Error> {
+                Ok(Signature::new_secp256k1(DUMMY_SIG.to_vec()))
+            }
+        }
+
+        let msg = Message {
+            to: Address::new_id(u64::arbitrary(g)),
+            from: Address::new_id(u64::arbitrary(g)),
+            ..Message::default()
+        };
```
Comment (lemmih) [inline vm/message/src/signed_message.rs:143]: Define as many fields as possible.

**Review by lemmih — CHANGES_REQUESTED** [review]

**Review by lemmih — APPROVED** [review]

File: vm/message/src/signed_message.rs [hunk 6]
```diff
@@ -124,6 +124,34 @@ impl Cbor for SignedMessage {
...
 
+#[cfg(test)]
+impl quickcheck::Arbitrary for SignedMessage {
+    fn arbitrary(g: &mut quickcheck::Gen) -> Self {
+        const DUMMY_SIG: [u8; 1] = [0u8];
+
+        struct DummySigner;
+        impl Signer for DummySigner {
+            fn sign_bytes(&self, _: &[u8], _: &Address) -> Result<Signature, anyhow::Error> {
+                Ok(Signature::new_secp256k1(DUMMY_SIG.to_vec()))
```
Comment (LesnyRumcajs) [inline vm/message/src/signed_message.rs:135]: it's used in one place only, I don't think there's a need to make it a const. Would `new_secp256k1(vec![0u8])` work? Not a big deal though.

**Pull request merged**